### ISSUE TITLE
Weird size issue with TextBox

### ISSUE THREAD

**Issue by mrj001** [AvaloniaUI/Avalonia#15187]
### Describe the bug

The TextBox is initially slightly too tall.  When I begin typing in it, the height adjusts to the correct value.

The weird part is that this does not happen if the `StackPanel.Styles` tag is removed from the following XAML, and it doesn't even apply any style to the TextBox.

### To Reproduce

The following XAML illustrates the problem.
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="400" d:DesignHeight="400"
        Width="400" Height="400"
        x:Class="TextBox.MainWindow"
        Title="TextBox">
  <StackPanel Orientation="Vertical" Margin="5">
    <StackPanel.Styles>
      <Style Selector="TextBlock">
        <Setter Property="Margin" Value="0,10,0,5" />
      </Style>
    </StackPanel.Styles>

    <TextBlock Text="A TextBox:" />
    <TextBox Name="TextBox" />
  </StackPanel>
</Window>

```

### Expected behavior

The initial size of the TextBox should be the same as after one starts typing in it.  ie. There should be no change in the vertical size of the TextBox when one begins typing in it.

### Avalonia version

11.0.10

### OS

Linux

**Comment by Gillibald** [comment]
You are styling the TextBox's watermark and expect unchanged measurements of the TextBox's template.

**Comment by mrj001** [comment]
@Gillibald, yes, that is what I am expecting.  You prompted me to review the [selector documentation](https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax#by-template), and I found this under "By Template" (italics added):
> You can match a control in a control template using the above syntax. _All the other selectors listed here work on the logical tree, but this selector can step into a template._

If I understand correctly, the TextBox's watermark is part of its visual tree, but not part of its logical tree.  Thus, my expectation remains that the TextBox should not change its size due to my style selector.
